WebTransport session should send origin header with the initial CONNECT request

AnkshitJain opened the pull request:
#### 2056a98ff3733785f9df63d751a4e5bea20d124b
<pre>
WebTransport session should send origin header with the initial CONNECT request
<a href="https://bugs.webkit.org/show_bug.cgi?id=284606">https://bugs.webkit.org/show_bug.cgi?id=284606</a>
<a href="https://rdar.apple.com/136263242">rdar://136263242</a>

Reviewed by Alex Christensen.

Adds the origin header to the CONNECT request as it is required for browser clients.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/cocoa/NetworkSoftLink.h: Added.
* Source/WebCore/PAL/pal/cocoa/NetworkSoftLink.mm: Added.
* Source/WebCore/PAL/pal/spi/cocoa/NetworkSPI.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::initializeWebTransportSession):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::initialize):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::createParameters):
(WebKit::NetworkTransportSession::initialize):
* Source/WebKit/WebProcess/Network/WebSocketProvider.cpp:
(WebKit::WebSocketProvider::initializeWebTransportSession):
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::initialize):
* Source/WebKit/WebProcess/Network/WebTransportSession.h:

Canonical link: <a href="https://commits.webkit.org/287827@main">https://commits.webkit.org/287827@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0ea8ce50fd4fe9e686004bdc77b200a1b3a4fb36

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81003 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/528 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34946 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85532 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31989 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83113 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/545 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8329 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63236 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/21005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84072 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/320 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73724 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43534 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/216 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27884 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30447 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/71766 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28434 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86967 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8233 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/5812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/71539 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8410 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69558 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70774 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14817 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/13740 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12564 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8194 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/13717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8031 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11551 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9839 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/PAL/pal/cocoa/NetworkSoftLink.h:
```diff
@@ -0,0 +1,38 @@
+
+SOFT_LINK_FRAMEWORK_FOR_HEADER(PAL, Network)
+
+SOFT_LINK_FUNCTION_FOR_HEADER(PAL, Network, nw_webtransport_options_add_connect_request_header, void, (nw_protocol_options_t options, const char* name, const char* value), (options, name, value))
```
Could we add a comment indicating this soft linking can be removed once rdar://??? is available on all tested OS builds?

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -46,11 +47,12 @@
+static RetainPtr<nw_parameters_t> createParameters(WebCore::SecurityOriginData&& origin)
 {
-    auto configureWebTransport = [](nw_protocol_options_t options) {
+    auto configureWebTransport = [&origin](nw_protocol_options_t options) {
```
This needs to be captured by value and not by reference because the stack that contains origin is likely gone by the time it is used.

achristensen07 commented on Source/WebKit/WebProcess/Network/WebSocketProvider.cpp:
```diff
@@ -53,7 +53,9 @@ void WebSocketProvider::initializeWebTransportSession(WebCore::ScriptExecutionCo
     ASSERT(is<Document>(context));
-    WebKit::WebTransportSession::initialize(url, WTFMove(completionHandler));
+    RefPtr<SecurityOrigin> origin = context.securityOrigin();
+    ASSERT(origin);
```
Could you put this before the is<WorkerGlobalScope> check above?  We will need it for both cases.  Also, let's null check the origin and return completionHandler(nullptr) if it is null just in case.

achristensen07 commented on Source/WebKit/WebProcess/Network/WebSocketProvider.cpp:
```diff
@@ -53,7 +53,9 @@ void WebSocketProvider::initializeWebTransportSession(WebCore::ScriptExecutionCo
-    WebKit::WebTransportSession::initialize(url, WTFMove(completionHandler));
+    RefPtr<SecurityOrigin> origin = context.securityOrigin();
+    ASSERT(origin);
+    WebKit::WebTransportSession::initialize(url, origin.get()->data(), WTFMove(completionHandler));
```
.get() is unnecessary because RefPtr has operator->

achristensen07 reviewed (changes requested):
Can the test server verify that the origin header was received?

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -46,11 +47,12 @@
+static RetainPtr<nw_parameters_t> createParameters(WebCore::SecurityOriginData&& origin)
 {
-    auto configureWebTransport = [](nw_protocol_options_t options) {
+    auto configureWebTransport = [origin = origin](nw_protocol_options_t options) {
```
origin = WTFMove(origin) prevents unnecessary copies.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
```diff
@@ -28,6 +28,7 @@
 #include "MessageReceiver.h"
 #include "MessageSender.h"
 #include <WebCore/ProcessQualified.h>
+#include <WebCore/SecurityOrigin.h>
```
Using a forward declaration will speed up build times:

namespace WebCore {
struct SecurityOriginData;
}

achristensen07 commented on Source/WebKit/WebProcess/Network/WebSocketProvider.cpp:
```diff
@@ -47,13 +47,21 @@ RefPtr<ThreadableWebSocketChannel> WebSocketProvider::createWebSocketChannel(Doc
 
 void WebSocketProvider::initializeWebTransportSession(WebCore::ScriptExecutionContext& context, const URL& url, CompletionHandler<void(RefPtr<WebCore::WebTransportSession>&&)>&& completionHandler)
 {
+    RefPtr<SecurityOrigin> origin = context.securityOrigin();
```
I think &lt;SecurityOrigin&gt; is unnecessary.

Commit:
WebTransport session should send origin header with the initial CONNECT request
https://bugs.webkit.org/show_bug.cgi?id=284606
rdar://136263242

Reviewed by Alex Christensen.

Adds the origin header to the CONNECT request as it is required for browser clients.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/cocoa/NetworkSoftLink.h: Added.
* Source/WebCore/PAL/pal/cocoa/NetworkSoftLink.mm: Added.
* Source/WebCore/PAL/pal/spi/cocoa/NetworkSPI.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::initializeWebTransportSession):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::initialize):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::createParameters):
(WebKit::NetworkTransportSession::initialize):
* Source/WebKit/WebProcess/Network/WebSocketProvider.cpp:
(WebKit::WebSocketProvider::initializeWebTransportSession):
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::initialize):
* Source/WebKit/WebProcess/Network/WebTransportSession.h:

Canonical link: https://commits.webkit.org/287827@main
